internal/controller/controllers/agent_controller.go

```diff
@@ -127,68 +123,71 @@ func (r *AgentReconciler) Reconcile(origCtx context.Context, req ctrl.Request) (
 		return ctrl.Result{}, nil
 	}
 
-	kubeKey := types.NamespacedName{
-		Namespace: agent.Spec.ClusterDeploymentName.Namespace,
-		Name:      agent.Spec.ClusterDeploymentName.Name,
-	}
-	clusterDeployment := &hivev1.ClusterDeployment{}
-
-	// Retrieve clusterDeployment
-	if err = r.Get(ctx, kubeKey, clusterDeployment); err != nil {
-		if k8serrors.IsNotFound(err) {
-			// Delete the agent, using a finalizer with pre-delete to deregister the host.
-			log.Infof("Cluster Deployment name: %s namespace: %s not found, deleting Agent",
-				agent.Spec.ClusterDeploymentName.Name, agent.Spec.ClusterDeploymentName.Namespace)
+	h, err := r.Installer.GetHostByKubeKey(req.NamespacedName)
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
 			return r.deleteAgent(ctx, log, req.NamespacedName)
+		} else {
+			log.WithError(err).Errorf("failed to retrieve Host %s from backend", agent.Name)
+			return ctrl.Result{RequeueAfter: defaultRequeueAfterOnError}, err
 		}
+	}
 
-		errMsg := fmt.Sprintf("failed to get clusterDeployment with name %s in namespace %s",
-			agent.Spec.ClusterDeploymentName.Name, agent.Spec.ClusterDeploymentName.Namespace)
-		log.WithError(err).Error(errMsg)
-		// Update that we failed to retrieve the clusterDeployment
-		return r.updateStatus(ctx, log, agent, nil, nil, errors.Wrapf(err, errMsg), !k8serrors.IsNotFound(err))
+	if agent.Spec.ClusterDeploymentName == nil && h.ClusterID != nil {
+		log.Debugf("ClusterDeploymentName is unset in Agent %s. unbind", agent.Name)
+		host, err := r.Installer.UnbindHostInternal(ctx, installer.UnbindHostParams{
+			HostID:     *h.ID,
+			InfraEnvID: h.InfraEnvID,
+		})
+		return r.updateStatus(ctx, log, agent, &host.Host, nil, err, !IsUserError(err))
 	}
 
-	// Retrieve cluster by ClusterDeploymentName from the database
-	cluster, err := r.Installer.GetClusterByKubeKey(kubeKey)
-	if err != nil {
-		if errors.Is(err, gorm.ErrRecordNotFound) {
-			// Delete the agent, using a finalizer with pre-delete to deregister the host.
-			log.Infof("Cluster name: %s namespace: %s not found in backend, deleting Agent",
-				agent.Spec.ClusterDeploymentName.Name, agent.Spec.ClusterDeploymentName.Namespace)
-			return r.deleteAgent(ctx, log, req.NamespacedName)
-		}
-		// Update that we failed to retrieve the cluster from the database
-		return r.updateStatus(ctx, log, agent, nil, nil, err, !errors.Is(err, gorm.ErrRecordNotFound))
+	if agent.Spec.ClusterDeploymentName != nil && h.ClusterID == nil{
```

Comment:
i think you can move it under the if agent.Spec.ClusterDeploymentName != nil at lne 150, since you will need to the clusterdeployment anyway, no?